diff --git a/src/Cedar/Console.c b/src/Cedar/Console.c
index 45bab426..481fdf8b 100644
--- a/src/Cedar/Console.c
+++ b/src/Cedar/Console.c
@@ -31,6 +31,9 @@
 #include <readline/readline.h>
 #include <readline/history.h>
 #endif
+#ifdef __ANDROID__
+#include <androidhelpers.h>
+#endif
 
 // Display the help for the command
 void PrintCmdHelp(CONSOLE *c, char *cmd_name, TOKEN_LIST *param_list)
@@ -2037,7 +2040,11 @@ wchar_t *Prompt(wchar_t *prompt_str)
 #else	// OS_WIN32
 	{
 		char *prompt = CopyUniToStr(prompt_str);
+		#ifdef __ANDROID__
+		char *s = strdup("");
+		#else
 		char *s = readline(prompt);
+		#endif
 		Free(prompt);
 
 		if (s != NULL)
@@ -2047,7 +2054,9 @@ wchar_t *Prompt(wchar_t *prompt_str)
 
 			if (IsEmptyStr(s) == false)
 			{
+				#ifndef __ANDROID__
 				add_history(s);
+				#endif
 			}
 
 			ret = CopyStrToUni(s);
diff --git a/src/Mayaqua/FileIO.c b/src/Mayaqua/FileIO.c
index 5e4b9c97..135b3a83 100644
--- a/src/Mayaqua/FileIO.c
+++ b/src/Mayaqua/FileIO.c
@@ -19,6 +19,9 @@
 #include "Win32.h"
 
 #include <Hamcore.h>
+#ifdef __ANDROID__
+#include <androidhelpers.h>
+#endif
 
 static char exe_file_name[MAX_SIZE] = "/tmp/a.out";
 static wchar_t exe_file_name_w[MAX_SIZE] = L"/tmp/a.out";
@@ -876,7 +879,10 @@ void GetExeDir(char *name, UINT size)
 	{
 		return;
 	}
-
+	#ifdef __ANDROID__
+	name = GetAndroidTmpDir();
+	size = StrLen(name);
+    #endif
 	GetDirNameFromFilePath(name, size, exe_file_name);
 }
 void GetExeDirW(wchar_t *name, UINT size)
@@ -916,6 +922,9 @@ void GetLogDir(char *name, UINT size)
 {
 #ifdef SE_LOGDIR
 	Format(name, size, SE_LOGDIR);
+#elif __ANDROID__
+	 name = GetAndroidLogDir();
+	 size = StrSize(name);
 #else
 	GetExeDir(name, size);
 #endif
@@ -934,6 +943,9 @@ void GetDbDir(char *name, UINT size)
 {
 #ifdef SE_DBDIR
 	Format(name, size, SE_DBDIR);
+#elif __ANDROID__
+	 name = GetAndroidDbDir();
+	 size = StrSize(name);
 #else
 	GetExeDir(name, size);
 #endif
@@ -952,6 +964,9 @@ void GetPidDir(char *name, UINT size)
 {
 #ifdef SE_PIDDIR
 	Format(name, size, SE_PIDDIR);
+#elif __ANDROID__
+	 name = GetAndroidTmpDir();
+	 size = StrSize(name);
 #else
 	GetExeDir(name, size);
 #endif
diff --git a/src/Mayaqua/Mayaqua.c b/src/Mayaqua/Mayaqua.c
index 86492751..135b0347 100644
--- a/src/Mayaqua/Mayaqua.c
+++ b/src/Mayaqua/Mayaqua.c
@@ -24,6 +24,9 @@
 
 #include <locale.h>
 #include <stdlib.h>
+#ifdef __ANDROID__
+#include <androidhelpers.h>
+#endif
 
 // Global variable
 bool g_memcheck;								// Enable memory check
@@ -621,12 +624,17 @@ void CheckUnixTempDir()
 		char tmp[128], tmp2[64];
 		UINT64 now = SystemTime64();
 		IO *o;
+#ifdef __ANDROID__
+			char* tmpDir = GetAndroidTmpDir();
+#else
+			char* tmpDir = "/tmp";
+#endif
 
-		MakeDir("/tmp");
+		MakeDir(tmpDir);
 
 		Format(tmp2, sizeof(tmp2), "%I64u", now);
 
-		Format(tmp, sizeof(tmp), "/tmp/.%s", tmp2);
+		Format(tmp, sizeof(tmp), "%s/.%s", tmpDir, tmp2);
 
 		o = FileCreate(tmp);
 		if (o == NULL)
@@ -634,7 +642,7 @@ void CheckUnixTempDir()
 			o = FileOpen(tmp, false);
 			if (o == NULL)
 			{
-				Print("Unable to use /tmp.\n\n");
+				Print("Unable to use temporary dir: (%s).\n\n", tmp);
 				exit(0);
 			}
 		}
diff --git a/src/Mayaqua/Unix.c b/src/Mayaqua/Unix.c
index 0c3778df..512afb9c 100755
--- a/src/Mayaqua/Unix.c
+++ b/src/Mayaqua/Unix.c
@@ -2728,6 +2728,7 @@ void UnixUsage(char *name)
 	UniPrint(_UU("UNIX_SVC_HELP"), svc_title, svc_name, svc_name, svc_title, svc_name, svc_title);
 }
 
+
 // Main function of the UNIX service
 UINT UnixService(int argc, char *argv[], char *name, SERVICE_FUNCTION *start, SERVICE_FUNCTION *stop)
 {
@@ -2737,6 +2738,14 @@ UINT UnixService(int argc, char *argv[], char *name, SERVICE_FUNCTION *start, SE
 		return 0;
 	}
 
+#ifdef __ANDROID__
+		// On Android, directly call UnixServiceMain without forking
+		if (argc >= 2 && StrCmpi(argv[1], UNIX_SVC_ARG_EXEC_SVC) == 0) {
+			UnixServiceMain(argc, argv, name, start, stop);
+		} else {
+			exit(EXIT_FAILURE);
+		}
+#else
 	if (argc >= 2 && StrCmpi(argv[1], UNIX_SVC_ARG_EXEC_SVC) == 0)
 	{
 		UINT pid;
@@ -2786,6 +2795,7 @@ RESTART_PROCESS:
 		// Start normally
 		UnixServiceMain(argc, argv, name, start, stop);
 	}
+#endif // __ANDROID__
 
 	return 0;
 }
diff --git a/src/vpnclient/vpncsvc.c b/src/vpnclient/vpncsvc.c
index 74d6da06..ca760655 100644
--- a/src/vpnclient/vpncsvc.c
+++ b/src/vpnclient/vpncsvc.c
@@ -43,4 +43,8 @@ int main(int argc, char *argv[])
 #endif	// OS_WIN32
 }
 
-
+#ifdef __ANDROID__
+extern int VpnClientMain(int argc, char *argv[]) {
+	return main(argc,argv);
+}
+#endif
